yt/utilities/on_demand_imports.py

```diff
@@ -22,6 +22,58 @@ def __call__(self, *args, **kwargs):
         raise self.error
 
 
+class OnDemandImport:
+    """
+    Implement a safely importable drop-in replacement for a module that
+    might not be installed, and fail *only* on the first attempt to use it.
+    """
+
+    def __init__(self, name: str):
+        if name.startswith("astropy"):
+            print(name)
+        self._name = name
+        self._module = None
+        self._submodules = {}
+
+    @property
```

Comment:
This would be a good usecase for `functools.cached_property`.

Reply:
which was my original implementation actually, but it's only available from Python 3.8 :)

Reply:
![image](https://user-images.githubusercontent.com/5411875/116892103-c4cad680-ac2f-11eb-996e-bb84a5e65bc1.png)